shell/platform/android/android_context_gl.cc

```diff
@@ -105,10 +106,119 @@ static bool TeardownContext(EGLDisplay display, EGLContext context) {
+ public:
+  void init(EGLDisplay display, EGLContext context) {
+    const char* extensions = eglQueryString(display, EGL_EXTENSIONS);
+
+    if (HasExtension(extensions, "EGL_KHR_partial_update")) {
+      set_damage_region = reinterpret_cast<PFNEGLSETDAMAGEREGIONKHRPROC>(
+          eglGetProcAddress("eglSetDamageRegionKHR"));
+    }
+
+    if (HasExtension(extensions, "EGL_EXT_swap_buffers_with_damage")) {
+      swap_buffers_with_damage =
+          reinterpret_cast<PFNEGLSWAPBUFFERSWITHDAMAGEEXTPROC>(
+              eglGetProcAddress("eglSwapBuffersWithDamageEXT"));
+    } else if (HasExtension(extensions, "EGL_KHR_swap_buffers_with_damage")) {
+      swap_buffers_with_damage =
+          reinterpret_cast<PFNEGLSWAPBUFFERSWITHDAMAGEEXTPROC>(
+              eglGetProcAddress("eglSwapBuffersWithDamageKHR"));
+    }
+
+    _partial_redraw_supported =
+        set_damage_region != nullptr && swap_buffers_with_damage != nullptr;
+  }
+
+  void SetDamageRegion(EGLDisplay display,
+                       EGLSurface surface,
+                       const std::optional<SkIRect>& region) {
+    if (set_damage_region && region) {
+      auto rects = RectToInts(display, surface, *region);
+      set_damage_region(display, surface, rects.data(), 1);
+    }
+  }
+
+  std::optional<SkIRect> InitialDamage(EGLDisplay display, EGLSurface surface) {
+    if (!_partial_redraw_supported) {
+      return std::nullopt;
+    }
+
+    EGLint age;
+    eglQuerySurface(display, surface, EGL_BUFFER_AGE_EXT, &age);
+
+    if (age == 0) {  // full repaint
+      return std::nullopt;
+    } else if (age == 1) {
+      // no initial damage
+      return SkIRect::MakeEmpty();
+    } else {
+      age -= 2;
+      SkIRect res;
+      for (auto i = damage_history_.begin();
+           i != damage_history_.end() && age >= 0; ++i, --age) {
+        res.join(*i);
+      }
+      return res;
+    }
+  }
+
+  bool SwapBuffersWithDamage(EGLDisplay display,
+                             EGLSurface surface,
+                             const std::optional<SkIRect>& damage) {
+    if (swap_buffers_with_damage && damage) {
+      damage_history_.push_back(*damage);
+      if (damage_history_.size() > 2) {
+        damage_history_.pop_front();
+      }
+      auto rects = RectToInts(display, surface, *damage);
+      return swap_buffers_with_damage(display, surface, rects.data(), 1);
+
+    } else {
+      return eglSwapBuffers(display, surface);
+    }
+  }
+
+ private:
+  std::vector<EGLint> RectToInts(EGLDisplay display,
+                                 EGLSurface surface,
+                                 const SkIRect& rect) {
+    std::vector<EGLint> res;
+    EGLint height;
+    eglQuerySurface(display, surface, EGL_HEIGHT, &height);
+
+    res.push_back(rect.left());
+    res.push_back(height - rect.bottom());
+    res.push_back(rect.width());
+    res.push_back(rect.height());
+
+    return res;
+  }
+
+  PFNEGLSETDAMAGEREGIONKHRPROC set_damage_region = nullptr;
```

Comment:
Class private members [end with underscore](https://google.github.io/styleguide/cppguide.html#Variable_Names). Consider renaming to `set_damage_region_`, `swap_buffers_with_damage_` and `partial_redraw_supported_`